### ISSUE TITLE
Normalize paths for 'per-file-ignores' - [closed]

### ISSUE THREAD

**Issue by asottile** [PyCQA/flake8#1213]
In GitLab by @ericvw on May 20, 2019, 20:37

_Merges per-file-ignores-normalize-path -> master_

Indicate that the `--per-file-ignores` option should be normalized to
ensure that the paths are relative to a detected configuration file's
location.

This also re-uses the `comma_separated_list` option customization to
parse out the independent values.  Eventually, the splitting of options
ends up in `utils.parse_comma_separated_list`, which also splits on
contiguous whitespace.  This behavior is asserted in the tests as well.

This is the minimal amount of change to address #517.

There are two things of note:

- Renaming `comma_separated_list` and related functions to be something
  like `separated_list` (to include comma separated and whitespace
  separated).  This renaming could be a follow-up commit.

- The exception message contains the normalized path and not the
  original value.  At the point where the exception is generated, there
  is nothing threaded through to know either the original and normalized
  value mapping or the normalized prefix to derive the original value for
  the error message.  This is a more significant refactoring to the
  options and option manager.

Closes: #517

**Comment by asottile** [comment]
In GitLab by @codecov on May 20, 2019, 20:39

# [Codecov](https://codecov.io/gl/pycqa/flake8/pull/321?src=pr&el=h1) Report
> Merging [#321](https://codecov.io/gl/pycqa/flake8/pull/321?src=pr&el=desc) into [master](https://codecov.io/gl/pycqa/flake8/commit/9b79c40cdb86b29cf65d7054b846da4a1b242430?src=pr&el=desc) will **not change** coverage.
> The diff coverage is `n/a`.

[![Impacted file tree graph](https://codecov.io/gl/pycqa/flake8/pull/321/graphs/tree.svg?width=650&token=QL6TVnaXih&height=150&src=pr)](https://codecov.io/gl/pycqa/flake8/pull/321?src=pr&el=tree)

```diff
@@           Coverage Diff           @@
##           master     #321   +/-   ##
=======================================
  Coverage   80.15%   80.15%           
=======================================
  Files          25       25           
  Lines        2363     2363           
  Branches      393      393           
=======================================
  Hits         1894     1894           
  Misses        389      389           
  Partials       80       80
```


| [Impacted Files](https://codecov.io/gl/pycqa/flake8/pull/321?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [src/flake8/main/options.py](https://codecov.io/gl/pycqa/flake8/pull/321/diff?src=pr&el=tree#diff-c3JjL2ZsYWtlOC9tYWluL29wdGlvbnMucHk=) | `100% <ø> (ø)` | :arrow_up: |

------

[Continue to review full report at Codecov](https://codecov.io/gl/pycqa/flake8/pull/321?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gl/pycqa/flake8/pull/321?src=pr&el=footer). Last update [9b79c40...1eada36](https://codecov.io/gl/pycqa/flake8/pull/321?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

**Comment by asottile** [comment]
In GitLab by @asottile on May 20, 2019, 20:47

:wave: hello friend, it's been quite a while!

my current plan was to rewrite the configuration parsing portion -- currently every setting is being parsed multiple times :S -- this would allow moving the `per-file-ignores` normalization to the configuration-loading portion (arguably where it belongs, this would also fix the exception normalization)

I don't think this patch is sufficient to fix cases like this:

```
per-file-ignores =
    file1,file2: A123
```

but I'll take it for a spin and see if there's something that can be improved here :thumbsup:

**Comment by asottile** [comment]
In GitLab by @ericvw on May 20, 2019, 21:05

> :wave: hello friend, it's been quite a while!

It has :wave:!

> my current plan was to rewrite the configuration parsing portion -- currently every setting is being parsed multiple times :S -- this would allow moving the `per-file-ignores` normalization to the configuration-loading portion

Yeah — I noticed that as well.  A rework of the configuration parsing where normalization and validation are conducted earlier seems like it will simplify a fair amount of interactions that occur elsewhere.  I held back on doing anything drastic without knowing what may have been in the pipeline to avoid duplication of effort.

If there is anything I can iterate upon, happy to do so.  If this patch is not sufficient enough, happy to either iterate on it or have it be dropped in favor of an alternative approach/path forward (i.e., rewrite of configuration parsing).

**Comment by asottile** [comment]
In GitLab by @asottile on May 20, 2019, 21:53

I haven't started on it yet beyond poking around and getting _some_ ideas on where I want to take it.  I recently added some type checking to hopefully make it slightly safer to refactor (before I plan to do major surgery I'd like to have some more sanity :laughing:)

My rough sketch plan is to have a rough set of config loading phases.  From my understanding this boils down to:

1. Initial commandline parsing ("prelim" parsing as the current code calls it
    - collect enough of the commandline in order to initialize plugins and configuration file loading
    - inputs: commandline
    - outputs: configuration loader / plugin loader
2. ask plugins to set up their options (plugin `.add_options(...)`)
3. parse commandline for real (may actually be after the next one)
    - use pwd as reference point for any file paths
4. parse config file(s) for real
    - TODO: should flake8 parse multiple config files? I think (?) the current approach is first-file-wins -- I like this it's simpler, need to double check
    - use configuration file as reference point for any file paths
    - merge over top commandline replacing values
5. apply extend_* options


This is honestly a bit rambly but I think most of the ideas are there.  I still need to do more investigation into exactly where the types are getting shuffled around and whether I'm missing major things.   This is just my "Current brain dump" before I have to catch some rest :laughing:

as above, I haven't started working on this and if it sounds like something you'd be interested in I'd definitely try and support you as best as possible -- it's not a small task and will likely need a lot of iteration and back and forth -- I'm happy to do it myself as well (as I was planning on doing over the next couple weekends)

**Comment by asottile** [comment]
In GitLab by @ericvw on May 27, 2019, 15:46

> I haven't started working on this and if it sounds like something you'd be interested in I'd definitely try and support you as best as possible -- it's not a small task and will likely need a lot of iteration and back and forth -- I'm happy to do it myself as well (as I was planning on doing over the next couple weekends)

That may be a bit too much than what I could reasonably commit to over the next couple of weeks.  If you end up breaking this down more concretely, I may be able to find some cycles to tackle some of those.

**Comment by asottile** [comment]
In GitLab by @asottile on May 28, 2019, 11:02

I think the smallest change to "fix" the problem is to only send the values through the normalization once (instead of repeatedly) -- this would allow the per-file-ignores parsing to happen only once and at the correct place (instead of being deferred)

Basically get this assertion to succeed:

```diff
diff --git a/src/flake8/utils.py b/src/flake8/utils.py
index f822388..9109728 100644
--- a/src/flake8/utils.py
+++ b/src/flake8/utils.py
@@ -39,6 +39,7 @@ def parse_comma_separated_list(value, regexp=COMMA_SEPARATED_LIST_RE):
     :rtype:
         list
     """
+    assert isinstance(value, (string_types, type(None))), value
     if not value:
         return []
 
```

**Comment by asottile** [comment]
In GitLab by @ericvw on May 28, 2019, 20:07

> Basically get this assertion to succeed

If I am understanding correctly, you want to change the contract of `parse_comma_separated_list()` to `# type: (str, Pattern[str]) -> List[str]`?

I'll take a stab at this.

**Comment by asottile** [comment]
In GitLab by @asottile on May 28, 2019, 20:44

I think it can actually be `Optional[str]` but yes essentially that :)

**Comment by asottile** [comment]
In GitLab by @ericvw on Aug 27, 2019, 13:49

closed

**Comment by asottile** [comment]
In GitLab by @ericvw on Aug 27, 2019, 13:55

Hey @asottile, I see that you are making some changes to the argument parsing already (e.g., moving to `argparse` via b66ebd7034090f96cb0806e5e2d8026b6e35d045).

I'm not sure entirely the direction you will be taking this, but being able to use `type` to `argparse` would be the place to conduct the configuration validation; however, I see that you are marking the parameter as deprecated.

**Comment by asottile** [comment]
In GitLab by @asottile on Aug 27, 2019, 13:58

yep `type` would be the way to do this, I only deprecated the `type='string'` optparse shim stuff -- it should be fine to do this in the parsing layer now :tada:

**Comment by asottile** [comment]
In GitLab by @ericvw on Aug 27, 2019, 14:02

Oh — I misunderstood what was deprecated!

I may take a crack at this again now that I better understand what's going on here.

**Comment by asottile** [comment]
In GitLab by @ericvw on Oct 1, 2019, 24:04

mentioned in merge request !358
